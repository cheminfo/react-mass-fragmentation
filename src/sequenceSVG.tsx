import { flushSync } from 'react-dom';
import { createRoot } from 'react-dom/client';

import { MassFragmentation } from './components/MassFragmentation';

export interface SequenceSVGOptions {
  /**
   * @default 50
   */
  leftRightBorders?: number;

  /**
   * @default 600
   */
  width?: number;

  /**
   * @default 30
   */
  spaceBetweenResidues?: number;

  /**
   * @default 12
   */
  spaceBetweenInternalLines?: number;

  /**
   * @default 2
   */
  strokeWidth?: number;

  /**
   * @default 'Verdana'
   */
  labelFontFamily?: string;

  /**
   * @default 8
   */
  labelSize?: number;

  /**
   * Sequence parsing options
   */
  parsing?: Record<string, any>;
  merge?: {
    /**
     * Merge results if only differs by a charge
     */
    charge?: boolean;
  };
  filter?: {
    /**
     * @default 0
     */
    minSimilarity?: number;

    /**
     * @default 0
     */
    minQuantity?: number;

    /**
     * minimal relative quantity. This value should be between 0 and 1 and supersede minQuantity.
     * @default 0
     */
    minRelativeQuantity?: number;

    /**
     * @default true
     */
    showInternals?: boolean;
  };
}

export function sequenceSVG(
  sequence: string,
  analysisResults,
  options: SequenceSVGOptions,
): string {
  const element = (
    <MassFragmentation
      sequence={sequence}
      analysisResults={analysisResults}
      {...options}
    />
  );
  const div = document.createElement('div');
  const root = createRoot(div);
  flushSync(() => {
    root.render(element);
  });
  return div.innerHTML;
}
