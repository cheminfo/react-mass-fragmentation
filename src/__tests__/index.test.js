import { test, expect } from 'vitest';

import { sequenceSVG } from '../sequenceSVG.tsx';

test('render: Molecule', () => {
  const sequence =
    '(Me)MQIFVKTLTGKT(OEt)IT(OMe)LEVEPSD(NH2)TIENVKAKIQD(NH2)KEGIPPDQQ(OMe)';

  const analysisResults = [
    { type: 'b38y33', similarity: 0.9012, charge: 3 },
    { type: 'b30y30', similarity: 0.8686, charge: 3 },
    { type: 'b40y40', similarity: 0.869, charge: 2 },
    { type: 'c1', similarity: 0.8919, charge: 3 },
    { type: 'c1', similarity: 0.9548, charge: 2 },
    { type: 'c5', similarity: 0.8764, charge: 2 },
    { type: 'x1', similarity: 0.8694, charge: 2 },
    { type: 'c28', similarity: 0.8463, charge: 2 },
    { type: 'c29', similarity: 0.8127, charge: 2 },
    { type: 'z28', similarity: 0.7976, charge: 2 },
    { type: 'c30', similarity: 0.7406, charge: 2 },
    { type: 'c27', similarity: 0.7157, charge: 2 },
    { type: 'c34', similarity: 0.6766, charge: 3 },
    { type: 'c17', similarity: 0.6617, charge: 2 },
    { type: 'z9', similarity: 0.6244, charge: 1 },
    { type: 'z9', similarity: 0.6164, charge: 2 },
    { type: 'z9', similarity: 0.5776, charge: 3 },
    { type: 'c38', similarity: 0.5044, charge: 2 },
  ];

  const options = {
    width: 700,
    leftRightBorders: 50,
    spaceBetweenResidues: 30,
    spaceBetweenInternalLines: 12,
    parsing: {
      kind: 'peptide',
    },
  };

  const svg = sequenceSVG(sequence, analysisResults, options);
  expect(svg).toHaveLength(13754);
  expect(svg).toMatchSnapshot();
});
