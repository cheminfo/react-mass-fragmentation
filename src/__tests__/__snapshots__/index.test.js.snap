// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render: Molecule 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 600 360"><g transform="translate(0 72)" font-family="Verdana" font-weight="bold"><g font-size="12"><text fill="#cccccc" x="50" text-anchor="middle">Me</text><text fill="#555555" x="80" text-anchor="middle">M</text><text fill="#555555" x="110" text-anchor="middle">Q</text><text fill="#555555" x="140" text-anchor="middle">I</text><text fill="#555555" x="170" text-anchor="middle">F</text><text fill="#555555" x="200" text-anchor="middle">V</text><text fill="#555555" x="230" text-anchor="middle">K</text><text fill="#555555" x="260" text-anchor="middle">T</text><text fill="#555555" x="290" text-anchor="middle">L</text><text fill="#555555" x="320" text-anchor="middle">T</text><text fill="#555555" x="350" text-anchor="middle">G</text><text fill="#555555" x="380" text-anchor="middle">K</text><text fill="darkviolet" x="410" text-anchor="middle">τ</text><text fill="#555555" x="440" text-anchor="middle">I</text><text fill="darkviolet" x="470" text-anchor="middle">¹τ</text><text fill="#555555" x="500" text-anchor="middle">L</text></g><g font-size="12" stroke-width="2" stroke-linecap="round" transform="translate(95)"><g transform="translate(362 0)"><line y2="-8" stroke="red" stroke-width="2"></line><line x2="5" y1="-8" y2="-13" stroke="red" stroke-width="2"></line><g transform="translate(8 -15)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">z28</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">79</text></g></g><g transform="translate(118 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c5</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">87</text></g></g><g transform="translate(-2 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="black" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c1</text><text fill="black" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="black" font-family="Verdana" font-size="4" font-weight="normal">95</text></g><g transform="translate(0 26)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c1</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">89</text></g></g></g><g transform="translate(0 -24)"><g transform="translate(80 0)"><line x2="600" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(276 -3)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b40y40</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">86</text></g></g><g transform="translate(380 -12)"><line x2="600" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(126 -3)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b30y30</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">86</text></g></g><g transform="translate(290 -24)"><line x2="600" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(171 -3)"><text fill="#333" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b38y33</text><text fill="#333" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#333" font-family="Verdana" font-size="4" font-weight="normal">90</text></g></g></g></g><g transform="translate(0 168)" font-family="Verdana" font-weight="bold"><g font-size="12"><text fill="#555555" x="50" text-anchor="middle">E</text><text fill="#555555" x="80" text-anchor="middle">V</text><text fill="#555555" x="110" text-anchor="middle">E</text><text fill="#555555" x="140" text-anchor="middle">P</text><text fill="#555555" x="170" text-anchor="middle">S</text><text fill="darkviolet" x="200" text-anchor="middle">δ</text><text fill="#555555" x="230" text-anchor="middle">T</text><text fill="#555555" x="260" text-anchor="middle">I</text><text fill="#555555" x="290" text-anchor="middle">E</text><text fill="#555555" x="320" text-anchor="middle">N</text><text fill="#555555" x="350" text-anchor="middle">V</text><text fill="#555555" x="380" text-anchor="middle">K</text><text fill="#555555" x="410" text-anchor="middle">A</text><text fill="#555555" x="440" text-anchor="middle">K</text><text fill="#555555" x="470" text-anchor="middle">I</text><text fill="#555555" x="500" text-anchor="middle">Q</text></g><g font-size="12" stroke-width="2" stroke-linecap="round" transform="translate(95)"><g transform="translate(-2 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c17</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">66</text></g></g><g transform="translate(298 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c27</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">71</text></g></g><g transform="translate(388 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c30</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">74</text></g></g><g transform="translate(358 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c29</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">81</text></g></g><g transform="translate(328 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c28</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">84</text></g></g></g><g transform="translate(0 -12)"><g transform="translate(0 0)"><line x2="600" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(316 -3)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b40y40</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">86</text></g></g><g transform="translate(0 -12)"><line x2="470" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(251 -3)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b30y30</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">86</text></g></g><g transform="translate(0 -24)"><line x2="600" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(316 -3)"><text fill="#333" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b38y33</text><text fill="#333" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#333" font-family="Verdana" font-size="4" font-weight="normal">90</text></g></g></g></g><g transform="translate(0 276)" font-family="Verdana" font-weight="bold"><g font-size="12"><text fill="darkviolet" x="50" text-anchor="middle">δ</text><text fill="#555555" x="80" text-anchor="middle">K</text><text fill="#555555" x="110" text-anchor="middle">E</text><text fill="#555555" x="140" text-anchor="middle">G</text><text fill="#555555" x="170" text-anchor="middle">I</text><text fill="#555555" x="200" text-anchor="middle">P</text><text fill="#555555" x="230" text-anchor="middle">P</text><text fill="#555555" x="260" text-anchor="middle">D</text><text fill="#555555" x="290" text-anchor="middle">Q</text><text fill="#555555" x="320" text-anchor="middle">Q</text><text fill="#cccccc" x="350" text-anchor="middle">3'</text></g><g font-size="12" stroke-width="2" stroke-linecap="round" transform="translate(95)"><g transform="translate(148 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c38</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">50</text></g></g><g transform="translate(-28 0)"><line y2="-8" stroke="red" stroke-width="2"></line><line x2="5" y1="-8" y2="-13" stroke="red" stroke-width="2"></line><g transform="translate(8 -15)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">z9</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">57</text></g><g transform="translate(8 -27)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">z9</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">61</text></g><g transform="translate(8 -39)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">z9</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+1</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">62</text></g></g><g transform="translate(28 0)"><line y2="-8" stroke="blue" stroke-width="2"></line><line x2="-5" y1="0" y2="5" stroke="blue" stroke-width="2"></line><g transform="translate(0 14)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">c34</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">67</text></g></g><g transform="translate(212 0)"><line y2="-8" stroke="red" stroke-width="2"></line><line x2="5" y1="-8" y2="-13" stroke="red" stroke-width="2"></line><g transform="translate(8 -15)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">x1</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">86</text></g></g></g><g transform="translate(0 -48)"><g transform="translate(0 0)"><line x2="290" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(161 -3)"><text fill="#999" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b40y40</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+2</text><text fill="#999" font-family="Verdana" font-size="4" font-weight="normal">86</text></g></g><g transform="translate(0 -12)"><line x2="230" stroke-linecap="round" stroke-width="2" stroke="orange"></line><g transform="translate(131 -3)"><text fill="#333" font-family="Verdana" font-size="8" font-weight="bold" text-anchor="end">b38y33</text><text fill="#333" font-family="Verdana" font-size="4" font-weight="normal" y="-4">+3</text><text fill="#333" font-family="Verdana" font-size="4" font-weight="normal">90</text></g></g></g></g><g transform="translate(50 312)"><g transform="translate(0 0)"><text fill="darkviolet" font-family="Verdana" font-size="8" font-weight="bold">τ = Thr(OEt)</text></g><g transform="translate(0 12)"><text fill="darkviolet" font-family="Verdana" font-size="8" font-weight="bold">¹τ = Thr(OMe)</text></g><g transform="translate(0 24)"><text fill="darkviolet" font-family="Verdana" font-size="8" font-weight="bold">δ = Asp(NH2)</text></g><g transform="translate(0 36)"><text fill="darkviolet" font-family="Verdana" font-size="8" font-weight="bold">3' = OMe</text></g></g></svg>"`;
